packages/assets-controllers/src/MultichainAssetsRatesController/MultichainAssetsRatesController.ts

```diff
@@ -332,75 +340,180 @@ export class MultichainAssetsRatesController extends StaticIntervalPollingContro
       if (!this.isActive) {
         return;
       }
-      const accounts = this.#listAccounts();
 
+      // Compute the set of unique assets from all accounts. It's important to
+      // deduplicate assets here to avoid duplicate requests to the Snap.
+      const accounts = this.#listAccounts();
+      const assetToSnapID = new Map<CaipAssetType, SnapId>();
       for (const account of accounts) {
-        const assets = this.#getAssetsForAccount(account.id);
-
-        if (assets?.length === 0) {
+        const snapId = account.metadata.snap?.id;
+        // Assets may still be updated if they exist in another account with a
+        // Snap ID.
+        //
+        // FIXME: Instead of using the Snap ID from the account, we should
+        // select the Snap based on the supported scopes defined in the Snaps'
+        // manifest.
+        if (snapId === undefined) {
           continue;
         }
 
-        const rates = await this.#getUpdatedRatesFor(account, assets);
-        // Apply these updated rates to controller state
-        this.#applyUpdatedRates(rates);
+        for (const asset of this.#getAssetsForAccount(account.id)) {
+          assetToSnapID.set(asset, snapId as SnapId);
```

Comment:
Maybe simpler to construct the Snap ID -> Asset List mapping here instead of constructing this mapping and then constructing the reverse mapping in the next function call?

Reply:
I did in two steps mainly because it also dedupes the assets. If I build the snapId -> assets[] mapping here I need to do another pass to remove the deplicates of each entry.